internal/framework/service/rulesets/model.go

```diff
@@ -14,22 +14,19 @@ type RulesetResourceModel struct {
 }
 
 type RulesModel struct {
-	Version                types.String                   `tfsdk:"version"`
 	Action                 types.String                   `tfsdk:"action"`
 	ActionParameters       []*ActionParametersModel       `tfsdk:"action_parameters"`
 	Description            types.String                   `tfsdk:"description"`
 	Enabled                types.Bool                     `tfsdk:"enabled"`
 	ExposedCredentialCheck []*ExposedCredentialCheckModel `tfsdk:"exposed_credential_check"`
 	Expression             types.String                   `tfsdk:"expression"`
 	ID                     types.String                   `tfsdk:"id"`
-	LastUpdated            types.String                   `tfsdk:"last_updated"`
 	Logging                []*LoggingModel                `tfsdk:"logging"`
 	Ratelimit              []*RatelimitModel              `tfsdk:"ratelimit"`
 	Ref                    types.String                   `tfsdk:"ref"`
 }
 
 type ActionParametersModel struct {
-	Version                  types.String                                 `tfsdk:"version"`
```

Comment:
from memory, the top level `version` is read only; this one is optionally provided. @zakcutner can confirm.

Reply:
yes, top level version is read only, when I tried to explicitly set it i have an error:
```
│ Error: Invalid Configuration for Read-Only Attribute
│ 
│   with cloudflare_ruleset.terraform_managed_resource,
│   on main.tf line 11, in resource "cloudflare_ruleset" "terraform_managed_resource":
│   11:     version     = "3"
│ 
│ Cannot set value for this attribute as the provider has marked it as read-only. Remove the configuration line setting the value.
│ 
│ Refer to the provider documentation or contact the provider developers for additional information about configurable and read-only attributes that are
│ supported.
```

And when I tried to set version for `action_parameters` I have also error, but different:
```
  # cloudflare_ruleset.terraform_managed_resource will be updated in-place
  ~ resource "cloudflare_ruleset" "terraform_managed_resource" {
        id      = "9662a2ab33a84458a1382122723bc411"
        name    = "inspire_firewall_custom"
        # (3 unchanged attributes hidden)

      ~ rules {
          ~ id           = "0834845279ee4892a27c36778ed48b88" -> (known after apply)
          ~ last_updated = "2024-01-26 15:05:38.398532 +0000 UTC" -> (known after apply)
          ~ ref          = "0834845279ee4892a27c36778ed48b88" -> (known after apply)
          ~ version      = "10" -> (known after apply)
            # (4 unchanged attributes hidden)
        }
      ~ rules {
          ~ id           = "b77ca99c7bef4ce2b902877e4fc4c80f" -> (known after apply)
          ~ last_updated = "2024-01-26 15:05:38.398532 +0000 UTC" -> (known after apply)
          ~ ref          = "b77ca99c7bef4ce2b902877e4fc4c80f" -> (known after apply)
          ~ version      = "1" -> (known after apply)
            # (4 unchanged attributes hidden)

          ~ action_parameters {
              + version = "3"
                # (1 unchanged attribute hidden)
            }

          ~ logging {
              ~ enabled = true -> false
            }
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
cloudflare_ruleset.terraform_managed_resource: Modifying... [id=9662a2ab33a84458a1382122723bc411]
╷
│ Error: error updating ruleset with ID "9662a2ab33a84458a1382122723bc411"
│ 
│   with cloudflare_ruleset.terraform_managed_resource,
│   on main.tf line 1, in resource "cloudflare_ruleset" "terraform_managed_resource":
│    1: resource "cloudflare_ruleset" "terraform_managed_resource" {
│ 
│ invalid JSON: unknown field "version"
```

Reply:
yep, that looks right. you're trying to use the `version` attribute in the wrong phase so you get that cryptic error message (there is an internal ticket to address that).